crates/settlement-clients/ethereum/src/conversion.rs

```diff
@@ -1,12 +1,304 @@
+pub(crate) fn vec_u8_32_to_vec_u256(slices: &[[u8; 32]]) -> EyreResult<Vec<U256>> {
     slices.iter().map(|slice| slice_u8_to_u256(slice)).collect()
 }
 
 /// Converts a `&[u8]` to `U256`.
-pub(crate) fn slice_u8_to_u256(slice: &[u8]) -> U256 {
-    U256::try_from_be_slice(slice).expect("could not convert u8 slice to U256")
+pub(crate) fn slice_u8_to_u256(slice: &[u8]) -> EyreResult<U256> {
+    U256::try_from_be_slice(slice).wrap_err_with(|| "could not convert &[u8] to U256".to_string())
+}
+
+// Function to convert a slice of u8 to a padded hex string
+// Function only takes a slice of length up to 32 elements
+// Pads the value on the right side with zeros only if the converted string has lesser than 64 characters.
+pub(crate) fn to_padded_hex(slice: &[u8]) -> String {
+    assert!(slice.len() <= 32, "Slice length must not exceed 32");
+    let hex = slice.iter().fold(String::new(), |mut output, byte| {
+        // 0: pads with zeros
+        // 2: specifies the minimum width (2 characters)
+        // x: formats the number as lowercase hexadecimal
+        // writes a byte value as a two-digit hexadecimal number (padded with a leading zero if necessary) to the specified output.
+        let _ = write!(output, "{byte:02x}");
+        output
+    });
+    format!("{:0<64}", hex)
+}
+
+/// Function to construct the transaction's `input data` for updating the state in the core contract.
+/// HEX Concatenation: MethodId, Offset, length for program_output, lines count, program_output, length for kzg_proof, kzg_proof
+/// All 64 chars, if lesser padded from left with 0s
+pub fn get_input_data_for_eip_4844(program_output: Vec<[u8; 32]>, kzg_proof: [u8; 48]) -> Result<String, Error> {
+    // bytes4(keccak256(bytes("updateStateKzgDA(uint256[],bytes)")))
+    let method_id_hex = "0xb72d42a1";
+
+    // offset for updateStateKzgDA is 64
+    let offset: u64 = 64;
+    let offset_hex = format!("{:0>64x}", offset);
+
+    // program_output
+    let program_output_length = program_output.len();
+    let program_output_hex = vec_u8_32_to_hex_string(program_output);
+
+    // length for program_output: 3*64 [offset, length, lines all have 64 char length] + length of program_output
+    let length_program_output = (3 * 64 + program_output_hex.len()) / 2;
+    let length_program_output_hex = format!("{:0>64x}", length_program_output);
+
+    // lines count for program_output
+    let lines_count_hex = format!("{:0>64x}", program_output_length);
+
+    // length of KZG proof
+    let length_kzg_hex = format!("{:0>64x}", kzg_proof.len());
+
+    // KZG proof
+    let kzg_proof_hex = u8_48_to_hex_string(kzg_proof);
+
+    let input_data = method_id_hex.to_string()
+        + &offset_hex
+        + &length_program_output_hex
+        + &lines_count_hex
+        + &program_output_hex
+        + &length_kzg_hex
+        + &kzg_proof_hex;
+
+    Ok(input_data)
+}
+
+pub(crate) fn vec_u8_32_to_hex_string(data: Vec<[u8; 32]>) -> String {
```

Comment:
Why do you need ownership here? Can't you pass a slice in all cases?

Reply:
Valid, improved. using ` &[[u8; 32]]` now